backend/src/halo2_impl.rs

```diff
@@ -11,6 +11,28 @@ impl<T: FieldElement> BackendImpl<T> for Halo2Prover {
         Halo2Prover::new(degree)
     }
 
+    fn add_verification_key(
+        &mut self,
+        pil: &Analyzed<T>,
+        fixed: &[(String, Vec<T>)],
+        vkey: Vec<u8>,
+    ) {
+        self.add_verification_key(pil, fixed, vkey)
+    }
+
+    fn verification_key(&self, pil: &Analyzed<T>, fixed: &[(String, Vec<T>)]) -> Vec<u8> {
+        self.verification_key(pil, fixed)
+    }
+
+    fn verify(
+        &self,
+        proof: &Proof,
+        verification_key: &[u8],
+        instances: &[Vec<T>],
+    ) -> Result<bool, String> {
+        self.verify(proof, verification_key, instances)
+    }
+
```

Comment:
We should really rethink how our backend is implemented. There's so much repetition. These functions are implemented 4 times in a row just calling each other. `Backend -> BackendImpl -> ConcreteBackend -> Halo2Impl -> Halo2`.

Reply:
To be honest I still don't even understand what's what in `lib.rs` because there are so many traits and impls. I don't think `BackendImplWithVerification` is needed because all proof systems have a verification key and verification, so it's just like `prove`.